### PR TITLE
8339771: RISC-V: Reduce icache flushes

### PR DESCRIPTION
**Pull request by robehn**
Hey, please consider,

All code which is offline (behind a barrier) do not need global icache flushes.
As we can instead in slow path locally (thread and hart) emit fence.i.
But if we were to be context switch do a hart which have not had fence.i emitted we can still fetch stale instructions.
To handle this case new now have kernel support:
https://docs.kernel.org/arch/riscv/cmodx.html

It's not perfect as we will be emitting fence.i on any context switch for any thread with this patch, even if that thread do not execute on code heap (non attached native thread), and even if there was no changes to code heap.
But this is in many cases much faster as the icache flush global IPI is very intrusive.
Particular cases are running a concurrent gc with small head room.
In such scenario I measured 15% increased throughput on VF2.
A large CPU or less head room (faster GC cycles) will yield even more performance boost.

Note that this requires 6.10 kernel.

I'm running VF2 with 6.11-rc3 kernel and this passed tier1-3. (With setting on)

Later we probably want this default on, but as it's hard to test I'll leave default off.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339771](https://bugs.openjdk.org/browse/JDK-8339771): RISC-V: Reduce icache flushes (**Enhancement** - P5)

### Reviewers
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer) Review applies to [5489a0d5](https://git.openjdk.org/jdk/pull/20913/files/5489a0d56d037df3f8ead3f7ee6b575e6cde322e)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20913/head:pull/20913` \
`$ git checkout pull/20913`

Update a local copy of the PR: \
`$ git checkout pull/20913` \
`$ git pull https://git.openjdk.org/jdk.git pull/20913/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20913`

View PR using the GUI difftool: \
`$ git pr show -t 20913`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20913.diff">https://git.openjdk.org/jdk/pull/20913.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20913#issuecomment-2338023934)

### CODE REVIEW

File: src/hotspot/os_cpu/linux_riscv/vm_version_linux_riscv.cpp [hunk 1]
```diff
@@ -102,6 +120,27 @@ void VM_Version::setup_cpu_available_features() {
   if (!RiscvHwprobe::probe_features()) {
     os_aux_features();
   }
+
+  // Linux kernel require Zifencei
+  if (!ext_Zifencei.enabled()) {
+    ext_Zifencei.enable_feature();
```
Comment (luhenry) [inline src/hotspot/os_cpu/linux_riscv/vm_version_linux_riscv.cpp:126]: That would deserve a `log_info` of sort, to make sure we can debug such state change in the wild

**Review by luhenry — APPROVED** [review]

**Comment by robehn** [comment]
Thank you @luhenry !

**Comment by RealFYang** [comment]
This is interesting. Thanks for trying this out.
BTW: Will this reflect on performance numbers of popular benchmarks workloads?

**Comment by RealFYang** [comment]
I am trying to compare this with https://github.com/openjdk/jdk/pull/9770 which removes unnecessary fence.i used in user space. I was thinking that this PR might add those fence.i back. But seems there are differences there. Previously, we emitted fence.i in `patch_callers_callsite()` and `MacroAssembler::emit_static_call_stub()`, which looks similar like what aarch64 does. But this PR adds fence.i in different places like in `generate_method_entry_barrier()` and `ZBarrierSetAssembler::patch_barrier_relocation()`. Do you have more details to help understand? Thanks.

**Comment by robehn** [comment]
```
default	-XX:-UseCtxFencei	-XX:+UseCtxFencei	
fop	2830	2731	3.63%
h2	25361	25403	-0.17%
jython	32482	32006	1.49%
luindex	4231	4369	-3.16%
lusearch	6142	5867	4.69%
lusearch-fix	6337	6243	1.51%
pmd	9171	8970	2.24%
			1.46%
zgc xmx512m			
fop	2583	2376	8.71%
h2	34520	33518	2.99%
jython	36494	35243	3.55%
luindex	4615	4497	2.62%
lusearch	6705	6732	-0.40%
lusearch-fix	6827	6485	5.27%
pmd	8385	8128	3.16%
			3.70%
zgc xmx128m			
fop	3336	2718	22.74%
jython	44385	35957	23.44%
luindex	4552	4460	2.06%
lusearch	9604	7272	32.07%
lusearch-fix	10095	7489	34.80%
pmd	11238	9169	22.57%
			22.95%
```

**Comment by robehn** [comment]
We have two major categories of cmodx.:

1.  Exectioner is not aware of any changes.
  1.1 Writer can icache_flush_all *or* executioner can *always* emit fence.i AND UseCtxFencei.
        If the code is not changed often the inline fence.i is just costly and icache_flush_all is better.
  1.2. We can enhance this by signaling a thread hand shake which will force all threads to emit a fence.i.
       Unclear if this is worth, because under some situations it can take a while to flush that out.
2. Exectioner is *aware* of any changes made:
  2.1 After safepoint. No need to do *icache_flush_all* in a safepoint. Just emit fence.i when leaving + UseCtxFencei.
  2.2 Nmethod entry barrier, same here.

The patch you are refering to was dealing with 1, which we shouldn't IMHO.
I'm dealing with 2, in a safepoint any code changes do not require *icache_flush_all* and all entries into nmethod is handle throught the barrier. Changes to nmethod do not need *icache_flush_all* of barrier used.

1: When we updates oops in code stream during safepoint or with nmethod barrier locked in Relocation::pd_set_data_value we do not need to flush.
2: When ZGC updates the color in code stream with nmethod barrier lock we do not need to flush.

**Comment by luhenry** [comment]
> ```
> default -XX:-UseCtxFencei -XX:+UseCtxFencei
> fop                  2830              2731  3.63%
> h2                  25361             25403 -0.17%
> jython              32482             32006  1.49%
> luindex              4231              4369 -3.16%
> lusearch             6142              5867  4.69%
> lusearch-fix         6337              6243  1.51%
> pmd                  9171              8970  2.24%
>                                              1.46%
> zgc xmx512m
> fop                  2583              2376  8.71%
> h2                  34520             33518  2.99%
> jython              36494             35243  3.55%
> luindex              4615              4497  2.62%
> lusearch             6705              6732 -0.40%
> lusearch-fix         6827              6485  5.27%
> pmd                  8385              8128  3.16%
>                                              3.70%
> zgc xmx128m
> fop                  3336              2718 22.74%
> jython              44385             35957 23.44%
> luindex              4552              4460  2.06%
> lusearch             9604              7272 32.07%
> lusearch-fix        10095              7489 34.80%
> pmd                 11238              9169 22.57%
>                                             22.95%
> ```

What's the unit of measurement? Also, lower is better? higher is better?

**Comment by robehn** [comment]
> What's the unit of measurement? Also, lower is better? higher is better?

Those are measured in milliseconds to completion, so lower is better.

**Review by RealFYang — COMMENTED** [review]
Thanks for sharing the performance numbers. It's a pity that I can't try this on my big RV machine for now which runs an older customized 6.1 kernel. Should we add back `verify_cross_modify_fence_not_required()` which was called in `MacroAssembler::read_polling_page()` and removed by https://github.com/openjdk/jdk/pull/9770?

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 2]
```diff
@@ -3054,6 +3054,13 @@ void MacroAssembler::membar(uint32_t order_constraint) {
   }
 }
 
+void MacroAssembler::cmodx_fence() {
+  BLOCK_COMMENT("cmodx fence");
+  if (VM_Version::supports_fencei_barrier()) {
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:3059]: The cmodx_fence() is called after a safepoint or thread local handshake.
Even if it costs a tens of cycles it is very little compared to hitting the poll, potentially comming from signal handler, hence therefore I kept it always 'on'.

But we are only guaranteed to have fence.i when running on Linux, this is code is not specific to linux.
I.e. MASM should be OS independent. As we already check this during boot in Linux specific code there is no way that assert can happen on Linux.

But maybe our JDK implementation should require fence.i ?
If we do then doing the assert make more sense, yes.

**Comment by robehn** [comment]
> Thanks for sharing the performance numbers. It's a pity that I can't try this on my big RV machine for now which runs an older customized 6.1 kernel. BTW: Should we add similar checking like `verify_cross_modify_fence_not_required()` which was called in `MacroAssembler::read_polling_page()` & `MacroAssembler::build_frame()` and removed by #9770? (-XX:+VerifyCrossModifyFence)

As we have two cases:
Code stream changed during a safepoint you must emit before leaving the safepoint, which means a thread in native return to Java must emit it. Polls are only disarmed by threads them self and they always do a cmodx_fence after a disarm.
This case is already covered by VerifyCrossModifyFence.

The second case, writing in the code stream with nmethod barrier locked methods, would be essentially be a verfication of the barrier and _patching_epoch.
If this needs verification, we need that regardless of this patch since do need both of them working, otherwise we may enter a nmethod with bad oop and we do need the loadload fence before entering otherwise we may load stale data.

So I don't believe this patch require any additional verification, but we may need verification.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (RealFYang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:3059]: Ah, I see. Make sense.

File: src/hotspot/cpu/riscv/relocInfo_riscv.cpp [hunk 3]
```diff
@@ -55,7 +55,10 @@ void Relocation::pd_set_data_value(address x, bool verify_only) {
       bytes = MacroAssembler::pd_patch_instruction_size(addr(), x);
       break;
   }
-  ICache::invalidate_range(addr(), bytes);
+
+  if (!UseCtxFencei) {
+    ICache::invalidate_range(addr(), bytes);
+  }
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/relocInfo_riscv.cpp:61]: One more question: Do we need a full fence (`OrderAccess::fence()`) here with `UseCtxFencei` after the patching? Like you do in `ZBarrierSetAssembler::patch_barrier_relocation()`:
```
  if (!UseCtxFencei) {
    // A full fence is generated before icache_flush by default in invalidate_word
    ICache::invalidate_range(addr, bytes);
  } else {
    OrderAccess::fence();
  }
```

Comment (robehn) [reply src/hotspot/cpu/riscv/relocInfo_riscv.cpp:61]: I actually didn't look at that.
As the old case did a full fence when calling ICache::invalidate_range(addr, bytes); and with that comment,
I assumed there was a reason for pointing it, so I just kept old behavior.
I'll see If I can figure out if/why/what.

Comment (robehn) [reply src/hotspot/cpu/riscv/relocInfo_riscv.cpp:61]: Sorry, now I realize what you are asking.

Comment (RealFYang) [reply src/hotspot/cpu/riscv/relocInfo_riscv.cpp:61]: The purpose is to to make a store to instruction memory visible to all RISC-V harts.
Check this:
```
static int icache_flush(address addr, int lines, int magic) {
  // To make a store to instruction memory visible to all RISC-V harts,
  // the writing hart has to execute a data FENCE before requesting that
  // all remote RISC-V harts execute a FENCE.I.

  // We need to make sure stores happens before the I/D cache synchronization.
  __asm__ volatile("fence rw, rw" : : : "memory");                                                     <==============

  RiscvFlushIcache::flush((uintptr_t)addr, ((uintptr_t)lines) << ICache::log2_line_size);

  return magic;
}
```

Comment (robehn) [reply src/hotspot/cpu/riscv/relocInfo_riscv.cpp:61]: Yes. The comment is a bit misleading.
A fence wr, wr is just for locally, on a hart, order instructions.
A fence wr wr do not flush a store from the store buffer, it just means that the hart must act such that the store appears to have happened before the syscall. Meaning the store can not happen after the syscall.

As we are not using the syscall, instead the other threads/harts must emit the fence.i them self, either by leaving a safepoint, hitting the patch_epoch_barrier and if they are moved to another hart.

Which means the store must happen before we leave the safepoint and disarming the nmethod barrier.
These already have store fences as we have a bunch of other stores which must be ordered.
For the nmethod barrier disarming is a full fence as the new epoch must happen before the disarm:

```
(nmethod barriers armed, implies storestore)
stores to instruction stream
store new patching epoch
storestore // the store to instruction and patching epoch must happen before disarm
disarm
```

Comment (robehn) [reply src/hotspot/cpu/riscv/relocInfo_riscv.cpp:61]: Code would in BarrierSetNMethod::set_guard_value(...)

Comment (RealFYang) [reply src/hotspot/cpu/riscv/relocInfo_riscv.cpp:61]: I see. Seems deserve a code comment. Also, does that mean the `OrderAccess::fence()` you added in `ZBarrierSetAssembler::patch_barrier_relocation()` unnecessary?

Comment (robehn) [reply src/hotspot/cpu/riscv/relocInfo_riscv.cpp:61]: Yes. I added due to the comment.
I'll remove it and update that comment and add new comment about above.

**Review by RealFYang — APPROVED** [review]
Updated change looks reasonable to me. Thanks for the answers and details.

**Review by Hamlin-Li — APPROVED** [review]

**Comment by robehn** [comment]
Thanks @Hamlin-Li, @luhenry and @RealFYang !
I'll let this sit until week, doing some extras test runs.

**Pull request closed without merging**